## 배열

### 배열 선언

### 일반적인 배열 선언

In [3]:
arr1 = [0, 0, 0, 0, 0, 0,]
arr1

[0, 0, 0, 0, 0, 0]

In [2]:
arr2 = [0] * 6
arr2

[0, 0, 0, 0, 0, 0]

#### 리스트 생성자 사용

In [ ]:
arr3 = list(range(6))       # list로 형변환
arr3

[0, 1, 2, 3, 4, 5]

#### 리스트 컴프리헨션

In [5]:
arr4 = [x for x in range(6)]
arr4

[0, 1, 2, 3, 4, 5]

In [10]:
# 어차피 사용하지 않는 변수라면 _(사용하지 않는 변수)로 변경하라
arr5 = [0 for _ in range(6)]
arr5

[0, 0, 0, 0, 0, 0]

In [12]:
arr4[0] + arr4[1] + arr4[2] + arr4[3] + arr4[4] + arr4[5]

15

### 2차원 배열

In [ ]:
# 선언
arr6 = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]   # 전체 3 x 4 배열
arr6

[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]

In [15]:
arr6[0]     # 0번 행 값들 전부

[1, 2, 3, 4]

In [16]:
arr6[0][3]  # 0행 3열 값

4

In [17]:
arr6[2][3] = 19
arr6

[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 19]]

### 리스트 컴프리헨션

In [19]:
arr7 = [[i+1]*4 for i in range(3)]
arr7

[[1, 1, 1, 1], [2, 2, 2, 2], [3, 3, 3, 3]]

In [21]:
arr7 = [[x for x in range(1 + y * 4, 5 + y * 4)] for y in range(3)]
arr7

[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]

In [22]:
[x for x in range(1,5)]

[1, 2, 3, 4]

### 배열 시간 복잡도

- 배열 요소에 접근하는 시간복잡도 : $ O (1) $. 소요시간이 전혀 걸리지 않음
- 배열에 맨 처음이나 중간에 삽입하는 시간 복잡도 : $ O (n) $
- 배열에 맨 처음이나 중간에 삭제하는 시간 복잡도 : $ O (n) $

#### 배열 선택시 고려할 점

- 할당할 메모리 크기 확인. 1차원 배열 최대 1000만개 2차원 배열 3000 * 3000 (900만개)
- 중간에 데이터 삽입이 많으면 시간복잡도가 높아짐

### 자주 사용하는 리스트 문법

#### append() 메서드로 데이터 추가

In [34]:
my_list = [1,2,3]
my_list.append(4)

In [24]:
my_list

[1, 2, 3, 4]

#### + 연산자 사용

In [35]:
my_list = my_list + [5,6]
my_list

[1, 2, 3, 4, 5, 6]

#### insert() 함수 사용

In [36]:
my_list.insert(3, 999)
my_list

[1, 2, 3, 999, 4, 5, 6]

#### pop() 함수로 데이터 추출

In [37]:
my_list.pop()

6

In [28]:
my_list

[1, 2, 3, 999, 4, 5]

In [38]:
my_list.pop(3)

999

In [33]:
my_list

[1, 2, 4, 5]

#### remove()로 데이터 삭제

In [ ]:
my_list.remove(3)   # list에서 3을 지운다

In [40]:
my_list

[1, 2, 4, 5]

#### 많이 사용하는 문법

In [41]:
fruits = ['apple', 'banana', 'cherry', 'apple', 'orange', 'banana', 'kiwi']

In [42]:
len(fruits)

7

In [43]:
fruits.index('banana')

1

In [46]:
fruits.index('banana', 2)

5

In [47]:
fruits.sort()
fruits

['apple', 'apple', 'banana', 'banana', 'cherry', 'kiwi', 'orange']

In [48]:
fruits.sort(reverse=True)
fruits

['orange', 'kiwi', 'cherry', 'banana', 'banana', 'apple', 'apple']

In [49]:
fruits.count('apple')

2

### 몸풀기 문제

#### 배열 정렬하기

|입력|출력|
|:--|:--|
|[1,-5,2,4,3]|[-5,1,2,4,3]|
|[2,1,1,3,2,5,4]|[1, 1, 2, 2, 3, 4, 5]|
|[6,1,7]|[1,6,7]|

In [6]:
def solution(arr):
    result = []
    return result

In [ ]:
# 기본 정렬
def solution(arr):
    arr.sort()
    return arr

In [ ]:
# arr1 = [1,-5,2,4,3]
arr1 = [2,1,1,3,2,5,4]
# arr1= [6,1,7]

solution(arr1)

[1, 1, 2, 2, 3, 4, 5]

In [7]:
# 원본은 보존하고 작업
def solution(arr):
    answer = sorted(arr, reverse=False)
    return answer

#### sort() 함수 사용않고 정렬알고리즘 구현
- $ O(n^2) $ 시간복잡도

In [ ]:
import time

def bubble_sort(arr):   # 버블정렬
    n = len(arr)
    for i in range(n):
        for j in range(n - i -1):   # 한번 돌고 나면 맨처음 요소는 뛰어넘고 정렬 계속
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]     # 파이썬에서 temp를 만들지 않고 swap 가능한 방법
    
    return arr

def do_sort(arr):   # sort() 사용
    arr.sort()
    return arr

def measure_time(func, arr):    # 시간 측정 함수
    start_time = time.time()    # 시작시간 저장
    result = func(arr)      # 함수 실행
    end_time = time.time()  # 함수 종료 시간 저장

    return end_time - start_time, result

arr =list(range(10000))

In [15]:
arr.sort(reverse=True)

##### 버블소트 시간측정

In [16]:
bubble_time, bubbel_result = measure_time(bubble_sort, arr)
print("첫 번째 코드 실행시간:", format(bubble_time, ".10f"))

첫 번째 코드 실행시간: 5.3593802452


In [18]:
len(bubbel_result)

10000

- 1~9999까지 정렬된 상태에서 버블소트 실행 1.6초 정도 소요
- 내림차순으로 역정렬된 상태에서 버블소트 실행 시 5.36초 소요

##### 내장 소트 시간측정

In [19]:
arr = list(range(10000))
arr.sort(reverse=True)

In [20]:
sort_time, sort_result = measure_time(do_sort, arr)

In [22]:
f"내장소트 실행시간 : {sort_time:.90f}"

'내장소트 실행시간 : 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000'

### 모의 테스트

- 공부할 때 사용하는 방법 VS Code. 실제 코테 환경에서는 불가능! 프로그래머스에서 바로 구현(디버깅 불가, 코드어시스턴스 사용불가)

#### 두개 뽑아서 더하기

- https://school.programmers.co.kr/learn/courses/30/lessons/68644

In [25]:
data = [5, 0, 2, 7]

In [32]:
def solution(numbers):
    answer = []     # 빈 배열이 기생성
    # 두 수를 선택해서 모든 경우의 수를 반복문으로 구함
    for i in range(len(numbers)):
        for j in range(i+1, len(numbers)):
            # i 인덱스와 j 인덱스의 값을 더해서 answer에 저장
            answer.append(numbers[i] + numbers[j])
    
    # 중복값 제거
    answer = sorted(set(answer))

    return answer

In [34]:
solution([2,1,3,4,1])

[2, 3, 4, 5, 6, 7]

#### 방문길이

- https://school.programmers.co.kr/learn/courses/30/lessons/49994

In [35]:
# 'ULURRDLLU'
# 'LULLLLLLU'

In [ ]:
# 좌표가 맵을 벗어나지 않았는지 확인하는 함수
def is_ok_in_map(nx, ny):
    # 맵의 좌표 범위는 0부터 10까지 (총 11칸)
    # 두 좌표(nx, ny)가 모두 0 이상 11 미만이면 True 반환
    return 0 <= nx < 11 and 0 <= ny < 11

# 방향(dir)에 따라 현재 좌표(x, y)를 이동시키는 함수
def move_position(dir, x, y):
    # dir 값에 따라 좌표를 이동시킴
    if dir.upper() == 'U':   # 위로 이동 (y 증가)
        nx, ny = x, y + 1
    elif dir.upper() == 'D': # 아래로 이동 (y 감소)
        nx, ny = x, y - 1
    elif dir.upper() == 'L': # 왼쪽으로 이동 (x 감소)
        nx, ny = x - 1, y
    elif dir.upper() == 'R': # 오른쪽으로 이동 (x 증가)
        nx, ny = x + 1, y
    return nx, ny

# 고유한 경로의 개수를 구하는 함수
def solution(dirs):
    x, y = 5, 5     # 시작 위치를 (5,5)로 설정 (중앙)
    answer = set()  # 이동 경로를 저장할 set (중복 제거용)

    # 주어진 방향 문자열을 하나씩 처리
    for dir in dirs:  # 예: 'LULLLLLLU'
        # 현재 위치(x, y)에서 방향(dir)으로 이동한 좌표 계산
        nx, ny = move_position(dir, x, y)

        # 이동한 좌표가 맵 범위 내에 있는 경우만 처리
        if is_ok_in_map(nx, ny):
            # (x, y) → (nx, ny) 이동 경로 저장
            answer.add((x, y, nx, ny))
            # (nx, ny) → (x, y)도 같은 경로로 간주하여 함께 저장
            answer.add((nx, ny, x, y))
            # 현재 위치를 이동한 위치로 업데이트
            x, y = nx, ny
        else:
            # 맵을 벗어나는 경우는 무시하고 다음 방향 처리
            continue

    # 양방향 경로가 저장되어 있으므로 전체 개수의 절반을 반환
    return len(answer) / 2  # 또는 정수 반환 원할 경우 // 2

In [37]:
solution('LULLLLLLU')

7.0